frappe/desk/form/meta.py

```diff
@@ -182,21 +183,38 @@ def add_custom_script(self):
 
 	def add_search_fields(self):
 		"""add search fields found in the doctypes indicated by link fields' options"""
+		# TODO: IF field is not found replace with useful message
 		for df in self.get("fields", {"fieldtype": "Link", "options": ["!=", "[Select]"]}):
 			if df.options:
-				search_fields = frappe.get_meta(df.options).search_fields
+				try:
+					search_fields = frappe.get_meta(df.options).search_fields
+				except frappe.DoesNotExistError:
+					self._show_missing_doctype_msg(df)
+
 				if search_fields:
 					search_fields = search_fields.split(",")
 					df.search_fields = [sf.strip() for sf in search_fields]
 
+	def _show_missing_doctype_msg(self, df):
+		# A link field is referring to non-existing doctype, this usually happens when
+		# customizations are removed or some custom app is removed but hasn't cleaned
+		# up after itself.
+		frappe.clear_last_message()
+		customize_form_link = f'<a href="/app/customize-form/?doc_type={self.name}">Customize Form</a>'
+		frappe.throw(
+			_(
+				"Field {0} is referring to non-existing doctype {1}, please remove the field from {2} or add the required doctype."
+			).format(frappe.bold(df.fieldname), frappe.bold(df.options), customize_form_link),
+			title="Missing DocType",
```

Comment:
This changing of exception also changes HTTP response code but IMO 404 didn't make sense for what was going on, the error is about incorrect/inconsistent doctype schema.